src/internal.rs

```diff
@@ -189,3 +189,140 @@ impl SizeLimit for Infinite {
+    
+    /// Serializes a sequence length.
+    fn serialize_len<W: Write, O: Options>(ser: &mut ::ser::Serializer<W, O>, len: usize) -> Result<()>;
+
+    /// Deserializes an enum discriminant.
+    fn deserialize_discriminant<'de, R: BincodeRead<'de>, O: Options>(ser: &mut ::de::Deserializer<R, O>) -> Result<u32>;
+
+    /// Deserializes a sequence length.
+    fn deserialize_len<'de, R: BincodeRead<'de>, O: Options>(ser: &mut ::de::Deserializer<R, O>) -> Result<usize>;
+}
+
+impl LengthEncoding for FixedLength {
+    #[inline(always)]
+    fn length_size(len: usize) -> u64 {
+        use std::mem::size_of_val;
+        size_of_val(&len) as u64
+    }
+
+    #[inline(always)]
+    fn discriminant_size(idx: u32) -> u64 {
+        use std::mem::size_of_val;
+        size_of_val(&idx) as u64
+    }
+
+    #[inline(always)]
+    fn serialize_discriminant<W: Write, O: Options>(ser: &mut ::ser::Serializer<W, O>, idx: u32) -> Result<()> {
+        use serde::Serialize;
+
+        idx.serialize(ser)
+    }
+
+    #[inline(always)]
+    fn serialize_len<W: Write, O: Options>(ser: &mut ::ser::Serializer<W, O>, len: usize) -> Result<()> {
+        use serde::Serialize;
+
+        (len as u64).serialize(ser)
+    }
+
+    fn deserialize_discriminant<'de, R: BincodeRead<'de>, O: Options>(de: &mut ::de::Deserializer<R, O>) -> Result<u32> {
+        serde::Deserialize::deserialize(de)
+    }
+
+    fn deserialize_len<'de, R: BincodeRead<'de>, O: Options>(de: &mut ::de::Deserializer<R, O>) -> Result<usize> {
+        serde::Deserialize::deserialize(de)
+    }
+}
+
+impl LengthEncoding for Varint {
+    #[inline(always)]
+    fn length_size(len: usize) -> u64 {
+        Varint::varint_size(len)
+    }
+
+    #[inline(always)]
+    fn discriminant_size(idx: u32) -> u64 {
+        Varint::varint_size(idx as usize)
+    }
+
+    #[inline(always)]
+    fn serialize_discriminant<W: Write, O: Options>(ser: &mut ::ser::Serializer<W, O>, idx: u32) -> Result<()> {
+        Varint::serialize_varint(ser, idx as usize)
+    }
+
+    #[inline(always)]
+    fn serialize_len<W: Write, O: Options>(ser: &mut ::ser::Serializer<W, O>, len: usize) -> Result<()> {
+        Varint::serialize_varint(ser, len)
+    }
+
+    fn deserialize_discriminant<'de, R: BincodeRead<'de>, O: Options>(de: &mut ::de::Deserializer<R, O>) -> Result<u32> {
+        Varint::deserialize_varint(de).map(|n| n as u32)
+    }
+
+    fn deserialize_len<'de, R: BincodeRead<'de>, O: Options>(de: &mut ::de::Deserializer<R, O>) -> Result<usize> {
+        Varint::deserialize_varint(de)
+    }
+}
+
+impl Varint {
+    fn varint_size(mut n: usize) -> u64 {
+        let mut bytes = 1;
+
+        while n > 127 {
+            n >>= 7;
+            bytes += 1;
+        }
+
+        bytes
+    }
+
+    fn serialize_varint<W: Write, O: Options>(ser: &mut ::ser::Serializer<W, O>, mut n: usize) -> Result<()> {
```

Comment:
This should take a `u64`, since in theory you could have sizes > the range of usize.